Adding support for chunked transfers

derjust opened the pull request:
Chunked transfers are not handled properly and time out while the transfer is still ongoing.

With this patchset, each messageChunk is considered a heart beat and resetting the timeout scheduler.

derjust committed:
Adding support for chunked transfers

Chunked transfers are not handled properly and time out while the
transfer is still ongoing.
With this patchset, each messageChunk is considered a heart beat and
resetting the timeout scheduler.

kristomasette commented on src/main/scala/org/caoilte/spray/routing/LogAccessRouting.scala:
```diff
@@ -55,7 +56,27 @@ class RequestAccessLogger(ctx: RequestContext, accessLogger: AccessLogger,
       accessLogger.logAccess(request, response, timeStampCalculator(Unit))
       forwardMsgAndStop(response)
     }
+    case Confirmed(ChunkedResponseStart(response), sentAck) => {
```
is it safe to assume these will always be Confirmed?

ie sent via responder ! MessageChunk.withAck(sentAck)?

kristomasette replied:
could re-write this case as:

case confirmed@Confirmed(ChunkResponseStart(response), _) => {
     chunkedResponse = response
     responder forward confirmed
}

derjust replied:
Chaned & covers now also un-ACKed

kristomasette commented on src/main/scala/org/caoilte/spray/routing/LogAccessRouting.scala:
```diff
@@ -55,7 +56,27 @@ class RequestAccessLogger(ctx: RequestContext, accessLogger: AccessLogger,
+
+      responder forward Confirmed(MessageChunk(data, extension), sentAck)
+    }
+    case response:ChunkedMessageEnd => {
```
ChunkMessageEnd could be Acked too.

derjust replied:
Chaned & covers now also un-ACKed

kristomasette commented on src/main/scala/org/caoilte/spray/routing/LogAccessRouting.scala:
```diff
@@ -55,7 +56,27 @@ class RequestAccessLogger(ctx: RequestContext, accessLogger: AccessLogger,
+      forwardMsgAndStop(response)
+    }
     case other => {
+      println(s"other $other")
```
did you mean to leave this println?

derjust replied:
Yes :)

kristomasette commented on src/main/scala/org/caoilte/spray/routing/LogAccessRouting.scala:
```diff
@@ -55,7 +56,27 @@ class RequestAccessLogger(ctx: RequestContext, accessLogger: AccessLogger,
+      responder forward Confirmed(ChunkedResponseStart(response), sentAck)
+    }
+    case Confirmed(MessageChunk(data, extension), sentAck) => {
+      // Each junk is considered a heart-beat: We are still producing output and the client is still consuming
```
junk == chunk

kristomasette replied:
might be able to re-write this

case confirmed@Confirmed(mc:MessageChunk,sentAck) => {
    //restart timeout thread
   responder forward confirmed
}

derjust replied:
Changed

derjust commented:
Added `withAck()` and `case`-handling for non-acked usage

kristomasette commented on src/main/scala/org/caoilte/spray/routing/LogAccessRouting.scala:
```diff
@@ -55,6 +62,26 @@ class RequestAccessLogger(ctx: RequestContext, accessLogger: AccessLogger,
+    case confirmed@Confirmed(ChunkedResponseStart(response), _) => handleChunkedResponseStart(confirmed, response)
+    case unconfirmed@ChunkedResponseStart(response) => handleChunkedResponseStart(unconfirmed, response)
+
+    case confirmed@(Confirmed(MessageChunk(_,_), _) | MessageChunk) => {
```
what about unconfirmed `MessageChunk`s?

kristomasette replied:
nice pattern match

kristomasette commented on src/main/scala/org/caoilte/spray/routing/LogAccessRouting.scala:
```diff
@@ -55,6 +62,26 @@ class RequestAccessLogger(ctx: RequestContext, accessLogger: AccessLogger,
       forwardMsgAndStop(response)
     }
+
+    case confirmed@Confirmed(ChunkedResponseStart(response), _) => handleChunkedResponseStart(confirmed, response)
```
Would this pattern match work?

```case confirmed@Confirmed(ChunkedResponseStart(response),_) | ChunkedResponseStart(response) => {
     chunkedResponse = response
     responder forward confirmed
}```

derjust replied:
Sadly no
```
LogAccessRouting.scala:69: illegal variable in pattern alternative
[error]     case confirmed@Confirmed(ChunkedResponseStart(response),_) | ChunkedResponseStart(response) => {
[error]                                                                                       ^
```

kristomasette replied:
Doh

kristomasette commented:
FWIW I'm :+1:

derjust committed:
Timeouts on chunked transfers were not handled properly

This commit adds a connection abort in case of a timeout happens
during a chunked transfer:
A reasonable tear down and/or sending any content to the client is
impossible as we have already started to transmit data.
Thus it is impossible to predict how a client would treat any 'appened'
error message to the stream anyway - and trailing headers are not
supported by any reasonable HTTP client library out there.

Therefore the connection is just aborted to get back to a "definied
state".

derjust committed:
Being a bit more conversative when it comes to
passing values back
